src/redux/actions/projectActions.ts

```diff
@@ -39,15 +43,34 @@ export default interface IProjectActions {
  * Dispatches Load Project action and resolves with IProject
  * @param project - Project to load
  */
-export function loadProject(project: IProject):
+export function loadProject(project: IProject, sharedToken?: ISecurityToken):
     (dispatch: Dispatch, getState: () => IApplicationState) => Promise<IProject> {
     return async (dispatch: Dispatch, getState: () => IApplicationState) => {
         const appState = getState();
         const projectService = new ProjectService();
 
+        let projectToken: ISecurityToken;
         // Lookup security token used to decrypt project settings
-        const projectToken = appState.appSettings.securityTokens
-            .find((securityToken) => securityToken.name === project.securityToken);
+        if (sharedToken) {
+            projectToken = sharedToken;
+            const existingToken = appState.appSettings.securityTokens.find((token) => token.name === projectToken.name);
+
+            if (!existingToken) {
+                // if we do not have project sharedToken, we need update security tokens in appState
+                dispatch(saveAppSettingsAction({
+                    securityTokens: [
+                        ...appState.appSettings.securityTokens,
+                        sharedToken
+                    ]
+                }));
+            } else if (existingToken.key !== sharedToken.key) {
+                const reason = interpolate(strings.shareProject.errors.tokenNameExist, {sharedTokenName: sharedToken.name})
+                toast.error(reason, { autoClose: false, closeOnClick: false });
+                return null;
+            }
+        } else {
+            projectToken = appState.appSettings.securityTokens.find((token) => token.name === project.securityToken);
```

Comment:
you already found the existing token, you don't need to find it again

Reply:
This is to find token for existing project (not shared).

Reply:
yeah, you get the same token on line 56, you could factor out and reduce redundant code. But not a big issue

Reply:
line 56 in we open shared string, if dont use shared string we are not getting inside this statement